### PR TITLE
MII dynamic update sample for work manager and data source tuning

### PR DESCRIPTION
**Pull request by alai8**
This PR is a WIP for merging into dynamicupdate branch/

It adds to the existing MII sample (plus its associated integration testing) to demonstrate dynamic on-line updates for:

- OWLS-85453:  work manager Maximum Threads Constraint and Minimum Threads Constraint updates
- OWLS-85452:  data source password and connection pool properties updates

It also updates the 'wl-pod-wait' script to wait for all pods to reach their expected 'introspect version' in addition to its current watch for 'restart version' and 'image'.

MII dynamic on-line updates rely on a new MII 'dynamic update' feature which can propagate configuration changes to a running domain using on-line calls without needing to roll/restart domain servers. This type of change requires both (a) updating the optional MII model config map and data source secret with the desired changes, plus (b) updating the domain resource introspectVersion and model 'online update' setting.

### CODE REVIEW

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml:134]: Shouldn't 'useOnlineUpdate' be under the 'model' stanza?  As far as I know it'll only be support for MII.

Comment (tbarnes-us) [reply kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml:134]: @jshum2479 ^^^

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml:131]: Suggested rerewording:

# If set to 'true' and an introspector job detects that
# WebLogic configuration changes for an already running
# domain are all dynamic, then the operator will attempt
# to use an online update to update WebLogic configuration
# instead of restarting all WebLogic server pods.

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/utils/patch-introspect-version.sh
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/model-in-image/utils/patch-introspect-version.sh:12]: rolling --> new.

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml
Comment (jshum2479) [reply kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml:134]: Still waiting for @rjeberhard  to have the schema finalized

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml
Comment (alai8) [reply kubernetes/samples/scripts/create-weblogic-domain/model-in-image/domain-resources/JRF/mii-update3-d1-JRF-v2-ds.yaml]: Thanks. updated.

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/utils/patch-introspect-version.sh
Comment (alai8) [reply kubernetes/samples/scripts/create-weblogic-domain/model-in-image/utils/patch-introspect-version.sh]: done

File: docs-source/content/samples/simple/domains/model-in-image/update4.md
Comment (alai8) [inline docs-source/content/samples/simple/domains/model-in-image/update4.md:24]: @jshum2479 We added the data source in a previous update, but we only want to update the SelfTuning configurations here. Do we need to keep the JDBCSystemResource section in the ConfigMap?

Comment (tbarnes-us) [reply docs-source/content/samples/simple/domains/model-in-image/update4.md:24]: I think it's helpful to layer changes on top of each-other, where teh first change was not dynamic - so keeping the JDBC system resource helps IMO.    On a related note, it'd be more clear/realistic if teh JDBC system resource remained in its original yaml file, and the updated configmap was built by adding both the original file and the new file together (instead of combining JDBC and WM tuning into the same file).

File: docs-source/content/samples/simple/domains/model-in-image/update3.md
Comment (tbarnes-us) [inline docs-source/content/samples/simple/domains/model-in-image/update3.md:420]: The alignment looks off - the new added output looks like it has a different indent than the original output. Does this look fine in hugo?

File: docs-source/content/samples/simple/domains/model-in-image/update4.md
Comment (tbarnes-us) [reply docs-source/content/samples/simple/domains/model-in-image/update4.md:24]: (I think you can have two "--from-file" arguments on the create cm command line.)

File: src/integration-tests/model-in-image/mii-sample-wrapper/generate-sample-doc.sh
Comment (tbarnes-us) [inline src/integration-tests/model-in-image/mii-sample-wrapper/generate-sample-doc.sh:174]: Purpose of this is to leave behind a utility shell script for creating the configmap - but it looks like update4 might overwrite the utility script generated by a previous update.   I could be wrong...

File: src/integration-tests/model-in-image/run-test.sh
Comment (tbarnes-us) [inline src/integration-tests/model-in-image/run-test.sh:454]: "updated datasources" --> "updated work managers"

File: docs-source/content/samples/simple/domains/model-in-image/update4.md
Comment (alai8) [reply docs-source/content/samples/simple/domains/model-in-image/update4.md:24]: Thanks for the suggestion! I have removed the datasource updates from the wm updates, and updated scripts to use two "--from-file" arguments when creating cm for update 4.

File: docs-source/content/samples/simple/domains/model-in-image/update3.md
Comment (alai8) [reply docs-source/content/samples/simple/domains/model-in-image/update3.md:420]: Thanks for catching this. Fixed.

File: src/integration-tests/model-in-image/mii-sample-wrapper/generate-sample-doc.sh
Comment (alai8) [reply src/integration-tests/model-in-image/mii-sample-wrapper/generate-sample-doc.sh:176]: generate-sample-doc.sh generates a separate *config.sh" for each update use case.

File: src/integration-tests/model-in-image/run-test.sh
Comment (alai8) [reply src/integration-tests/model-in-image/run-test.sh]: Fixed.

File: docs-source/content/samples/simple/domains/model-in-image/update2.md
Comment (tbarnes-us) [inline docs-source/content/samples/simple/domains/model-in-image/update2.md:465]: Alignment?

File: kubernetes/samples/scripts/create-weblogic-domain/model-in-image/model-configmaps/workmanager/model.20.workmanager.yaml
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/model-in-image/model-configmaps/workmanager/model.20.workmanager.yaml:12]: I thought the JDBC SR was going to be removed from here?  (Since the original yaml for the JDBC SR is now loaded into the new configmap too...)

Comment (alai8) [reply kubernetes/samples/scripts/create-weblogic-domain/model-in-image/model-configmaps/workmanager/model.20.workmanager.yaml]: Sorry, missed this file in previous commit. Should be fixed now. Thanks!

File: docs-source/content/samples/simple/domains/model-in-image/update2.md
Comment (alai8) [reply docs-source/content/samples/simple/domains/model-in-image/update2.md:465]: Alignment should be fixed now. Thanks

File: docs-source/content/samples/simple/domains/model-in-image/update4.md
Comment (jshum2479) [reply docs-source/content/samples/simple/domains/model-in-image/update4.md:24]: The merged model at any time is the source of truth (i.e. model in the image and configmap are merged into a single model).  The delta is based on the differences between the current running and the newly merged model and then used for update.  So, always add to, remove, update to the configmap to form a final model that you want.  If you remove everything in the configmap and just add a new wm, then all elements in the model will be removed and ending with just the wm.

Comment (tbarnes-us) [reply docs-source/content/samples/simple/domains/model-in-image/update4.md:24]: I think we're all on the same page. Resolving. Feel free to unresolve if you disagree.

**Comment by tbarnes-us** [comment]
@jshum2479 I am taking over this branch and this pull request from Anthony in order to add sample on-line changes for a data source on top of Anthony's sample on-line changes for a work-manager.  Please do not merge until the new work is added and reviewed.  Thanks!

**Comment by jshum2479** [comment]
ok, my branch has the updated schema though.

**Comment by alai8** [comment]
> ok, my branch has the updated schema though.

Thanks. I've updated this branch to include your latest schema, and made changes to the new sample accordingly.

**Review by rosemarymarano — COMMENTED** [review]
I reviewed and edited the files as seen in this commit. @tbarnes-us I'll post the changes.txt file in Slack for you.

**Review by rosemarymarano — APPROVED** [review]

**Review by tbarnes-us — APPROVED** [review]

**Review by jshum2479 — APPROVED** [review]
LGTM

**Comment by rjeberhard** [comment]
@alai8, the "WIP" in the title usually means "work in progress." Is that still the case for this change? Note: you don't need my permission to merge this PR because the target branch isn't a protected branch, like develop or master.

**Comment by tbarnes-us** [comment]
> @alai8, the "WIP" in the title usually means "work in progress." Is that still the case for this change? Note: you don't need my permission to merge this PR because the target branch isn't a protected branch, like develop or master.

@mriccell is planning on reviewing before the merge.

**Pull request merged**
